### PR TITLE
Add attribute form model base test coverage

### PR DESCRIPTION
**Pull request by kaustuvpokharel**
The attribute form model base was basically untested, so the old test only poked at 4 things through the derived model, none of the constraint/visibility/sync logic that actually runs the form so this adds sections for hard + soft constraints, constraint descriptions, tab layout, field sync across tabs, container visibility, and delete.

two calls I made that are worth a look:
- constraint tests build their own throwaway layer instead of reusing the fixture. the fixture sets editable = "str" is null on str, and constraints only evaluate when the field's editable, so putting a constraint there passes for the wrong reason (field goes readonly, validity forced true). own layer keeps it honest.
- visibility checks rowCount() instead of reading the CurrentlyVisible role, since the proxy just filters hidden tabs out, same way tst_featureForm already does it

### CODE REVIEW

**Comment by qfield-fairy** [comment]
### 🍎 MacOS DMG universal builds
Download a [MacOS DMG universal build of this PR for testing](https://nightly.link/opengisch/QField/actions/runs/33269153343/qfield-dmg.zip).
*(Built from commit fb5d1a446562634ee98d2219023dcc2982171b37)*




### 📱 Android builds
Download an [Android arm64 build of this PR for testing](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-arm64.zip).
*(Built from commit fb5d1a446562634ee98d2219023dcc2982171b37)*
<details>
<summary>Other Android architectures</summary>

- [arm-neon](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-arm-neon.zip)
- [x64](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-x64.zip)
- [x86](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-x86.zip)
- [**all access arm64**](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield_all_access-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-arm64.zip)
- [all access arm-neon](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield_all_access-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-arm-neon.zip)
- [all access x64](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield_all_access-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-x64.zip)
- [all access x86](https://nightly.link/opengisch/QField/actions/runs/33269153334/qfield_all_access-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-android-x86.zip)
</details>




### 🪟 Windows builds
Download a [Windows build of this PR for testing](https://nightly.link/opengisch/QField/actions/runs/33269153290/qfield-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-windows-x64.zip) and for [arm64](https://nightly.link/opengisch/QField/actions/runs/33269153290/qfield-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-windows-arm64.zip).

*(Built from commit fb5d1a446562634ee98d2219023dcc2982171b37)*




### 🐧 Linux AppImage builds
Download a [Linux AppImage build of this PR for testing](https://nightly.link/opengisch/QField/actions/runs/33269153402/qfield-dev-7892-57ccce0228d69dda2e5b8be94934e4b653f5b8ce-linux-x64.AppImage.zip).
*(Built from commit fb5d1a446562634ee98d2219023dcc2982171b37)*

File: test/test_attributeformmodel.cpp [hunk 1]
```diff
@@ -100,6 +111,191 @@ TEST_CASE( "AttributeFormModel" )
...
+  SECTION( "HardConstraint" )
+  {
+    std::unique_ptr<QgsVectorLayer> clean = std::make_unique<QgsVectorLayer>( QStringLiteral( "Point?crs=EPSG:3857&field=fid:integer&field=val:string" ), QStringLiteral( "Clean Layer" ), QStringLiteral( "memory" ) );
+    REQUIRE( clean->isValid() );
+    clean->setFieldConstraint( 1, QgsFieldConstraints::ConstraintNotNull, QgsFieldConstraints::ConstraintStrengthHard );
+
+    std::unique_ptr<QfAttributeFormModel> model = std::make_unique<QfAttributeFormModel>();
+    std::unique_ptr<QfFeatureModel> fModel = std::make_unique<QfFeatureModel>();
+    model->setFeatureModel( fModel.get() );
+    fModel->setCurrentLayer( clean.get() );
```
Comment (nirvn) [inline test/test_attributeformmodel.cpp]: Let's use clearer variable names here. model -> formModel, fModel -> featureModel. That'll help others understand the code more easily.

File: test/test_attributeformmodel.cpp [hunk 2]
```diff
@@ -100,6 +111,191 @@ TEST_CASE( "AttributeFormModel" )
...
+    layer->setFieldConstraint( 1, QgsFieldConstraints::ConstraintNotNull, QgsFieldConstraints::ConstraintStrengthHard );
+
+    std::unique_ptr<QfAttributeFormModel> model = std::make_unique<QfAttributeFormModel>();
+    std::unique_ptr<QfFeatureModel> fModel = std::make_unique<QfFeatureModel>();
+    model->setFeatureModel( fModel.get() );
+    fModel->setCurrentLayer( layer.get() );
+
+    const QModelIndex strField = indexForField( model.get(), 1 );
+    REQUIRE( strField.isValid() );
+    REQUIRE( !model->data( strField, QfAttributeFormModel::ConstraintDescription ).toString().isEmpty() );
```
Comment (nirvn) [inline test/test_attributeformmodel.cpp]: Let's do even better and compare against an expected value. We know that already.

**Review by nirvn — APPROVED** [review]
Thanks.
